// @superset-ui/chart-controls/esm/components/ControlHeader.js
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { t } from '@superset-ui/core';
import { InfoTooltipWithTrigger } from './InfoTooltipWithTrigger';
import { Tooltip } from './Tooltip';import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";



















export default function ControlHeader({
  name,
  description,
  label,
  tooltipOnClick,
  onClick,
  warning,
  danger,
  leftNode,
  rightNode,
  validationErrors = [],
  renderTrigger = false,
  hovered = false,
  required = false
}) {
  const renderOptionalIcons = () => {
    if (hovered) {
      return (
        _jsxs("span", { children: [
          description &&
          _jsxs("span", { children: [
            _jsx(InfoTooltipWithTrigger, {
              label: t('description'),
              tooltip: description,
              placement: "top",
              onClick: tooltipOnClick }
            ), ' '] }
          ),

          renderTrigger &&
          _jsxs("span", { children: [
            _jsx(InfoTooltipWithTrigger, {
              label: t('bolt'),
              tooltip: t('Changing this control takes effect instantly'),
              placement: "top",
              icon: "bolt" }
            ), ' '] }
          )] }

        ));

    }
    return null;
  };

  if (!label) {
    return null;
  }
  const labelClass = validationErrors.length > 0 ? 'text-danger' : '';
  return (
    _jsxs("div", { className: "ControlHeader", children: [
      _jsx("div", { className: "pull-left", children:
        _jsxs("label", { className: "control-label", htmlFor: name, children: [
          leftNode && _jsx("span", { children: leftNode }),
          _jsx("span", {
            role: "button",
            tabIndex: 0,
            onClick: onClick,
            className: labelClass,
            style: { cursor: onClick ? 'pointer' : '' }, children:

            label }
          ), ' ',
          warning &&
          _jsxs("span", { children: [
            _jsx(Tooltip, { id: "error-tooltip", placement: "top", title: warning, children:
              _jsx("i", { className: "fa fa-exclamation-circle text-warning" }) }
            ), ' '] }
          ),

          danger &&
          _jsxs("span", { children: [
            _jsx(Tooltip, { id: "error-tooltip", placement: "top", title: danger, children:
              _jsx("i", { className: "fa fa-exclamation-circle text-danger" }) }
            ), ' '] }
          ),

          validationErrors.length > 0 &&
          _jsxs("span", { children: [
            _jsx(Tooltip, {
              id: "error-tooltip",
              placement: "top",
              title: validationErrors.join(' '), children:

              _jsx("i", { className: "fa fa-exclamation-circle text-danger" }) }
            ), ' '] }
          ),

          renderOptionalIcons(),
          required &&
          _jsx("span", { className: "text-danger m-l-4", children:
            _jsx("strong", { children: "*" }) }
          )] }

        ) }
      ),
      rightNode && _jsx("div", { className: "pull-right", children: rightNode }),
      _jsx("div", { className: "clearfix" })] }
    ));

}
