// @superset-ui/chart-controls/esm/shared-controls/customControls.js
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
  ContributionType,
  ensureIsArray,
  GenericDataType,
  getColumnLabel,
  getMetricLabel,


  t } from
'@superset-ui/core';
import {



  isDataset } from
'../types';
import { isTemporalColumn } from '../utils';
import {
  DEFAULT_XAXIS_SORT_SERIES_DATA,
  SORT_SERIES_CHOICES } from
'../constants';
import { checkColumnType } from '../utils/checkColumnType';
import { isSortable } from '../utils/isSortable';

export const contributionModeControl = {
  name: 'contributionMode',
  config: {
    type: 'SelectControl',
    label: t('Contribution Mode'),
    default: null,
    choices: [
    [null, t('None')],
    [ContributionType.Row, t('Row')],
    [ContributionType.Column, t('Series')]],

    description: t('Calculate contribution per series or row')
  }
};

const xAxisSortVisibility = ({ controls }) => {var _controls$groupby, _controls$metrics;return (
    isSortable(controls) &&
    ensureIsArray(controls == null ? void 0 : (_controls$groupby = controls.groupby) == null ? void 0 : _controls$groupby.value).length === 0 &&
    ensureIsArray(controls == null ? void 0 : (_controls$metrics = controls.metrics) == null ? void 0 : _controls$metrics.value).length === 1);};

const xAxisMultiSortVisibility = ({
  controls


}) => {var _controls$groupby2, _controls$metrics2;return (
    isSortable(controls) && (
    !!ensureIsArray(controls == null ? void 0 : (_controls$groupby2 = controls.groupby) == null ? void 0 : _controls$groupby2.value).length ||
    ensureIsArray(controls == null ? void 0 : (_controls$metrics2 = controls.metrics) == null ? void 0 : _controls$metrics2.value).length > 1));};

export const xAxisSortControl = {
  name: 'x_axis_sort',
  config: {
    type: 'XAxisSortControl',
    label: (state) => {var _state$form_data;return (
        ((_state$form_data = state.form_data) == null ? void 0 : _state$form_data.orientation) === 'horizontal' ?
        t('Y-Axis Sort By') :
        t('X-Axis Sort By'));},
    description: t('Decides which column to sort the base axis by.'),
    shouldMapStateToProps: () => true,
    mapStateToProps: (state, controlState) => {var _controls$x_axis, _controls$metrics3, _controls$timeseries_, _controls$x_axis2, _controls$datasource;
      const { controls, datasource } = state;
      const dataset = isDataset(datasource) ? datasource : undefined;
      const columns = [controls == null ? void 0 : (_controls$x_axis = controls.x_axis) == null ? void 0 : _controls$x_axis.value].filter(
        Boolean
      );
      const metrics = [
      ...ensureIsArray(controls == null ? void 0 : (_controls$metrics3 = controls.metrics) == null ? void 0 : _controls$metrics3.value),
      controls == null ? void 0 : (_controls$timeseries_ = controls.timeseries_limit_metric) == null ? void 0 : _controls$timeseries_.value].
      filter(Boolean);
      const metricLabels = [...new Set(metrics.map(getMetricLabel))];
      const options = [
      ...columns.map((column) => {var _dataset$verbose_map;
        const value = getColumnLabel(column);
        return {
          value,
          label: (dataset == null ? void 0 : (_dataset$verbose_map = dataset.verbose_map) == null ? void 0 : _dataset$verbose_map[value]) || value
        };
      }),
      ...metricLabels.map((value) => {var _dataset$verbose_map2;return {
          value,
          label: (dataset == null ? void 0 : (_dataset$verbose_map2 = dataset.verbose_map) == null ? void 0 : _dataset$verbose_map2[value]) || value
        };})];


      const shouldReset = !(
      typeof controlState.value === 'string' &&
      options.map((option) => option.value).includes(controlState.value) &&
      !isTemporalColumn(
        getColumnLabel(controls == null ? void 0 : (_controls$x_axis2 = controls.x_axis) == null ? void 0 : _controls$x_axis2.value),
        controls == null ? void 0 : (_controls$datasource = controls.datasource) == null ? void 0 : _controls$datasource.datasource
      ));


      return {
        shouldReset,
        options
      };
    },
    visibility: xAxisSortVisibility
  }
};

export const xAxisSortAscControl = {
  name: 'x_axis_sort_asc',
  config: {
    type: 'CheckboxControl',
    label: (state) => {var _state$form_data2;return (
        ((_state$form_data2 = state.form_data) == null ? void 0 : _state$form_data2.orientation) === 'horizontal' ?
        t('Y-Axis Sort Ascending') :
        t('X-Axis Sort Ascending'));},
    default: true,
    description: t('Whether to sort ascending or descending on the base Axis.'),
    visibility: ({ controls }) => {var _controls$x_axis_sort;return (
        (controls == null ? void 0 : (_controls$x_axis_sort = controls.x_axis_sort) == null ? void 0 : _controls$x_axis_sort.value) !== undefined &&
        xAxisSortVisibility({ controls }));}
  }
};

export const xAxisForceCategoricalControl = {
  name: 'xAxisForceCategorical',
  config: {
    type: 'CheckboxControl',
    label: () => t('Force categorical'),
    default: false,
    description: t('Treat values as categorical.'),
    initialValue: (control, state) => {var _state$form_data3;return (
        (state == null ? void 0 : (_state$form_data3 = state.form_data) == null ? void 0 : _state$form_data3.x_axis_sort) !== undefined || control.value);},
    renderTrigger: true,
    visibility: ({ controls }) => {var _controls$x_axis3, _controls$datasource2;return (
        checkColumnType(
          getColumnLabel(controls == null ? void 0 : (_controls$x_axis3 = controls.x_axis) == null ? void 0 : _controls$x_axis3.value),
          controls == null ? void 0 : (_controls$datasource2 = controls.datasource) == null ? void 0 : _controls$datasource2.datasource,
          [GenericDataType.Numeric]
        ));},
    shouldMapStateToProps: () => true
  }
};

export const xAxisSortSeriesControl = {
  name: 'x_axis_sort_series',
  config: {
    type: 'SelectControl',
    freeForm: false,
    label: (state) => {var _state$form_data4;return (
        ((_state$form_data4 = state.form_data) == null ? void 0 : _state$form_data4.orientation) === 'horizontal' ?
        t('Y-Axis Sort By') :
        t('X-Axis Sort By'));},
    choices: SORT_SERIES_CHOICES,
    default: DEFAULT_XAXIS_SORT_SERIES_DATA.sort_series_type,
    renderTrigger: true,
    description: t('Decides which measure to sort the base axis by.'),
    visibility: xAxisMultiSortVisibility
  }
};

export const xAxisSortSeriesAscendingControl = {
  name: 'x_axis_sort_series_ascending',
  config: {
    type: 'CheckboxControl',
    label: (state) => {var _state$form_data5;return (
        ((_state$form_data5 = state.form_data) == null ? void 0 : _state$form_data5.orientation) === 'horizontal' ?
        t('Y-Axis Sort Ascending') :
        t('X-Axis Sort Ascending'));},
    default: DEFAULT_XAXIS_SORT_SERIES_DATA.sort_series_ascending,
    description: t('Whether to sort ascending or descending on the base Axis.'),
    renderTrigger: true,
    visibility: ({ controls }) => {var _controls$x_axis_sort2;return (
        (controls == null ? void 0 : (_controls$x_axis_sort2 = controls.x_axis_sort_series) == null ? void 0 : _controls$x_axis_sort2.value) !== undefined &&
        xAxisMultiSortVisibility({ controls }));}
  }
};
