// @superset-ui/chart-controls/lib/components/ColumnOption.js
(function () {var enterModule = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.enterModule : undefined;enterModule && enterModule(module);})();var __signature__ = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.default.signature : function (a) {return a;}; /**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { useState, useLayoutEffect } from 'react';
import { css, SafeMarkdown, styled } from '@superset-ui/core';
import { Tooltip } from './Tooltip';
import { ColumnTypeLabel } from './ColumnTypeLabel/ColumnTypeLabel';
import CertifiedIconWithTooltip from './CertifiedIconWithTooltip';

import {
  getColumnLabelText,
  getColumnTooltipNode,
  getColumnTypeTooltipNode } from
'./labelUtils';
import { SQLPopover } from './SQLPopover';
import InfoTooltipWithTrigger from './InfoTooltipWithTrigger';import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";







const StyleOverrides = styled.span`
  display: flex;
  align-items: center;
  svg {
    margin-right: ${({ theme }) => theme.gridUnit}px;
  }
`;

export function ColumnOption({
  column,
  labelRef,
  showType = false
}) {
  const { expression, column_name, type_generic } = column;
  const hasExpression = expression && expression !== column_name;
  const warningMarkdown =
  column.warning_markdown || column.warning_text || column.error_text;
  const type = hasExpression ? 'expression' : type_generic;
  const [tooltipText, setTooltipText] = useState(column.column_name);
  const [columnTypeTooltipText, setcolumnTypeTooltipText] = useState(
    column.type
  );

  useLayoutEffect(() => {
    setTooltipText(getColumnTooltipNode(column, labelRef));
    setcolumnTypeTooltipText(getColumnTypeTooltipNode(column));
  }, [labelRef, column]);

  return (
    _jsxs(StyleOverrides, { children: [
      showType && type !== undefined &&
      _jsx(Tooltip, {
        id: "metric-type-tooltip",
        title: columnTypeTooltipText,
        placement: "bottomRight",
        align: { offset: [8, -2] }, children:

        _jsx("span", { children:
          _jsx(ColumnTypeLabel, { type: type }) }
        ) }
      ),

      _jsx(Tooltip, { id: "metric-name-tooltip", title: tooltipText, children:
        _jsx("span", {
          className: "option-label column-option-label",
          css: (theme) => css`
            margin-right: ${theme.gridUnit}px;
          `,
          ref: labelRef, children:

          getColumnLabelText(column) }
        ) }
      ),
      hasExpression && _jsx(SQLPopover, { sqlExpression: expression }),
      column.is_certified &&
      _jsx(CertifiedIconWithTooltip, {
        metricName: column.metric_name,
        certifiedBy: column.certified_by,
        details: column.certification_details }
      ),

      warningMarkdown &&
      _jsx(InfoTooltipWithTrigger, {
        className: "text-warning",
        icon: "warning",
        tooltip: _jsx(SafeMarkdown, { source: warningMarkdown }),
        label: `warn-${column.column_name}`,
        iconsStyle: { marginLeft: 0 }, ...(
        column.error_text && {
          className: 'text-danger',
          icon: 'exclamation-circle'
        }) }
      )] }

    ));

}__signature__(ColumnOption, "useState{[tooltipText, setTooltipText](column.column_name)}\nuseState{[columnTypeTooltipText, setcolumnTypeTooltipText](column.type)}\nuseLayoutEffect{}");const _default =

ColumnOption;export default _default;;(function () {var reactHotLoader = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.default : undefined;if (!reactHotLoader) {return;}reactHotLoader.register(StyleOverrides, "StyleOverrides", "/Users/evan_1/GitHub/superset/superset-frontend/packages/superset-ui-chart-controls/src/components/ColumnOption.tsx");reactHotLoader.register(ColumnOption, "ColumnOption", "/Users/evan_1/GitHub/superset/superset-frontend/packages/superset-ui-chart-controls/src/components/ColumnOption.tsx");reactHotLoader.register(_default, "default", "/Users/evan_1/GitHub/superset/superset-frontend/packages/superset-ui-chart-controls/src/components/ColumnOption.tsx");})();;(function () {var leaveModule = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.leaveModule : undefined;leaveModule && leaveModule(module);})();
