// @superset-ui/chart-controls/lib/components/labelUtils.jsx
import { css, styled, t } from '@superset-ui/core';
const TooltipSectionWrapper = styled.div `
  ${({ theme }) => css `
    display: flex;
    flex-direction: column;
    font-size: ${theme.typography.sizes.s}px;
    line-height: 1.2;

    &:not(:last-of-type) {
      margin-bottom: ${theme.gridUnit * 2}px;
    }
    &:last-of-type {
      display: -webkit-box;
      -webkit-line-clamp: 40;
      -webkit-box-orient: vertical;
      overflow: hidden;
      text-overflow: ellipsis;
    }
  `}
`;
const TooltipSectionLabel = styled.span `
  ${({ theme }) => css `
    font-weight: ${theme.typography.weights.bold};
  `}
`;
const TooltipSection = ({ label, text, }) => (<TooltipSectionWrapper>
    <TooltipSectionLabel>{label}</TooltipSectionLabel>: <span>{text}</span>
  </TooltipSectionWrapper>);
export const isLabelTruncated = (labelRef) => !!(labelRef?.current?.scrollWidth > labelRef?.current?.clientWidth);
export const getColumnLabelText = (column) => column.verbose_name || column.column_name;
export const getColumnTypeTooltipNode = (column) => {
    if (!column.type) {
        return null;
    }
    return <TooltipSection label={t('Column type')} text={column.type}/>;
};
export const getColumnTooltipNode = (column, labelRef) => {
    if ((!column.column_name || !column.verbose_name) &&
        !column.description &&
        !isLabelTruncated(labelRef)) {
        return null;
    }
    return (<>
      {column.column_name && (<TooltipSection label={t('Column name')} text={column.column_name}/>)}
      {column.verbose_name && (<TooltipSection label={t('Label')} text={column.verbose_name}/>)}
      {column.description && (<TooltipSection label={t('Description')} text={column.description}/>)}
    </>);
};
export const getMetricTooltipNode = (metric, labelRef) => {
    if (!metric.verbose_name &&
        !metric.description &&
        !metric.label &&
        !isLabelTruncated(labelRef)) {
        return null;
    }
    return (<>
      <TooltipSection label={t('Metric name')} text={metric.metric_name}/>
      {(metric.label || metric.verbose_name) && (<TooltipSection label={t('Label')} text={metric.label || metric.verbose_name}/>)}
      {metric.description && (<TooltipSection label={t('Description')} text={metric.description}/>)}
    </>);
};
