// @superset-ui/chart-controls/lib/components/MetricOption.js
(function () {var enterModule = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.enterModule : undefined;enterModule && enterModule(module);})();var __signature__ = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.default.signature : function (a) {return a;}; /**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { useState, useLayoutEffect } from 'react';

import {
  css,
  styled,

  SafeMarkdown } from

'@superset-ui/core';
import InfoTooltipWithTrigger from './InfoTooltipWithTrigger';
import { ColumnTypeLabel } from './ColumnTypeLabel/ColumnTypeLabel';
import CertifiedIconWithTooltip from './CertifiedIconWithTooltip';
import Tooltip from './Tooltip';
import { getMetricTooltipNode } from './labelUtils';
import { SQLPopover } from './SQLPopover';import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";

const FlexRowContainer = styled.div`
  align-items: center;
  display: flex;

  > svg {
    margin-right: ${({ theme }) => theme.gridUnit}px;
  }
`;











export function MetricOption({
  metric,
  labelRef,
  openInNewWindow = false,
  showFormula = true,
  showType = false,
  shouldShowTooltip = true,
  url = ''
}) {
  const verbose = metric.verbose_name || metric.metric_name || metric.label;
  const link = url ?
  _jsx("a", { href: url, target: openInNewWindow ? '_blank' : '', rel: "noreferrer", children:
    verbose }
  ) :

  verbose;


  const label =
  _jsx("span", {
    className: "option-label metric-option-label",
    css: (theme) => css`
        margin-right: ${theme.gridUnit}px;
      `,
    ref: labelRef, children:

    link }
  );


  const warningMarkdown =
  metric.warning_markdown || metric.warning_text || metric.error_text;

  const [tooltipText, setTooltipText] = useState(metric.metric_name);

  useLayoutEffect(() => {
    setTooltipText(getMetricTooltipNode(metric, labelRef));
  }, [labelRef, metric]);

  return (
    _jsxs(FlexRowContainer, { className: "metric-option", children: [
      showType && _jsx(ColumnTypeLabel, { type: "expression" }),
      shouldShowTooltip ?
      _jsx(Tooltip, { id: "metric-name-tooltip", title: tooltipText, children:
        label }
      ) :

      label,

      showFormula && metric.expression &&
      _jsx(SQLPopover, { sqlExpression: metric.expression }),

      metric.is_certified &&
      _jsx(CertifiedIconWithTooltip, {
        metricName: metric.metric_name,
        certifiedBy: metric.certified_by,
        details: metric.certification_details }
      ),

      warningMarkdown &&
      _jsx(InfoTooltipWithTrigger, {
        className: "text-warning",
        icon: "warning",
        tooltip: _jsx(SafeMarkdown, { source: warningMarkdown }),
        label: `warn-${metric.metric_name}`,
        iconsStyle: { marginLeft: 0 }, ...(
        metric.error_text && {
          className: 'text-danger',
          icon: 'exclamation-circle'
        }) }
      )] }

    ));

}__signature__(MetricOption, "useState{[tooltipText, setTooltipText](metric.metric_name)}\nuseLayoutEffect{}");;(function () {var reactHotLoader = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.default : undefined;if (!reactHotLoader) {return;}reactHotLoader.register(FlexRowContainer, "FlexRowContainer", "/Users/evan_1/GitHub/superset/superset-frontend/packages/superset-ui-chart-controls/src/components/MetricOption.tsx");reactHotLoader.register(MetricOption, "MetricOption", "/Users/evan_1/GitHub/superset/superset-frontend/packages/superset-ui-chart-controls/src/components/MetricOption.tsx");})();;(function () {var leaveModule = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.leaveModule : undefined;leaveModule && leaveModule(module);})();
