// @superset-ui/chart-controls/lib/components/SQLPopover.js
(function () {var enterModule = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.enterModule : undefined;enterModule && enterModule(module);})();var __signature__ = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.default.signature : function (a) {return a;}; /**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { useEffect, useState } from 'react';
import { Popover } from 'antd';


import { CalculatorOutlined } from '@ant-design/icons';
import { css, styled, useTheme, t } from '@superset-ui/core';import { jsx as _jsx } from "@emotion/react/jsx-runtime";

const StyledCalculatorIcon = styled(CalculatorOutlined)`
  ${({ theme }) => css`
    color: ${theme.colors.grayscale.base};
    font-size: ${theme.typography.sizes.s}px;
    & svg {
      margin-left: ${theme.gridUnit}px;
      margin-right: ${theme.gridUnit}px;
    }
  `}
`;

export const SQLPopover = (props) => {
  const theme = useTheme();
  const [AceEditor, setAceEditor] = useState(null);
  useEffect(() => {
    Promise.all([
    import('react-ace'),
    import('ace-builds/src-min-noconflict/mode-sql')]
    ).then(([reactAceModule]) => {
      setAceEditor(() => reactAceModule.default);
    });
  }, []);

  if (!AceEditor) {
    return null;
  }
  return (
    _jsx(Popover, {
      content:
      _jsx(AceEditor, {
        mode: "sql",
        value: props.sqlExpression,
        editorProps: { $blockScrolling: true },
        setOptions: {
          highlightActiveLine: false,
          highlightGutterLine: false
        },
        minLines: 2,
        maxLines: 6,
        readOnly: true,
        wrapEnabled: true,
        style: {
          border: `1px solid ${theme.colors.grayscale.light2}`,
          background: theme.colors.secondary.light5,
          maxWidth: theme.gridUnit * 100
        } }
      ),

      placement: "bottomLeft",
      arrowPointAtCenter: true,
      title: t('SQL expression'), ...
      props, children:

      _jsx(StyledCalculatorIcon, {}) }
    ));

};__signature__(SQLPopover, "useTheme{theme}\nuseState{[AceEditor, setAceEditor](null)}\nuseEffect{}", () => [useTheme]);;(function () {var reactHotLoader = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.default : undefined;if (!reactHotLoader) {return;}reactHotLoader.register(StyledCalculatorIcon, "StyledCalculatorIcon", "/Users/evan_1/GitHub/superset/superset-frontend/packages/superset-ui-chart-controls/src/components/SQLPopover.tsx");reactHotLoader.register(SQLPopover, "SQLPopover", "/Users/evan_1/GitHub/superset/superset-frontend/packages/superset-ui-chart-controls/src/components/SQLPopover.tsx");})();;(function () {var leaveModule = typeof reactHotLoaderGlobal !== 'undefined' ? reactHotLoaderGlobal.leaveModule : undefined;leaveModule && leaveModule(module);})();
