// @boost/terminal/lib/cursor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ansi_escapes_1 = __importDefault(require("ansi-escapes"));
exports.default = {
    backward: ansi_escapes_1.default.cursorBackward,
    down: ansi_escapes_1.default.cursorDown,
    forward: ansi_escapes_1.default.cursorForward,
    hide: ansi_escapes_1.default.cursorHide,
    nextLine: ansi_escapes_1.default.cursorNextLine,
    position: ansi_escapes_1.default.cursorGetPosition,
    prevLine: ansi_escapes_1.default.cursorPrevLine,
    restorePosition: ansi_escapes_1.default.cursorRestorePosition,
    savePosition: ansi_escapes_1.default.cursorSavePosition,
    show: ansi_escapes_1.default.cursorShow,
    startLine: ansi_escapes_1.default.cursorLeft,
    to: ansi_escapes_1.default.cursorMove,
    toAbsolute: ansi_escapes_1.default.cursorTo,
    up: ansi_escapes_1.default.cursorUp,
};
