// @superset-ui/chart-controls/lib/shared-controls/components/RadioButtonControl.jsx
import { useTheme } from '@superset-ui/core';
import ControlHeader from '../../components/ControlHeader';
export default function RadioButtonControl({ value: initialValue, options, onChange, ...props }) {
    const currentValue = initialValue || options[0][0];
    const theme = useTheme();
    return (<div css={{
            '.btn svg': {
                position: 'relative',
                top: '0.2em',
            },
            '.btn:focus': {
                outline: 'none',
            },
            '.control-label': {
                color: theme.colors.grayscale.base,
                marginBottom: theme.gridUnit,
            },
            '.control-label + .btn-group': {
                marginTop: '1px',
            },
            '.btn-group .btn-default': {
                color: theme.colors.grayscale.dark1,
            },
            '.btn-group .btn.active': {
                background: theme.colors.grayscale.light4,
                fontWeight: theme.typography.weights.bold,
                boxShadow: 'none',
            },
        }}>
      <ControlHeader {...props}/>
      <div className="btn-group btn-group-sm">
        {options.map(([val, label]) => (<button key={JSON.stringify(val)} type="button" className={`btn btn-default ${val === currentValue ? 'active' : ''}`} onClick={() => {
                onChange(val);
            }}>
            {label}
          </button>))}
      </div>
    </div>);
}
