// @superset-ui/chart-controls/lib/types.js
export const isStandardizedFormData = (formData) => formData?.standardizedFormData?.controls &&
    formData?.standardizedFormData?.memorizedFormData &&
    Array.isArray(formData.standardizedFormData.controls.metrics) &&
    Array.isArray(formData.standardizedFormData.controls.columns);
// Ref:
//  - superset-frontend/src/explore/components/ConditionalFormattingControl.tsx
export var Comparator;
(function (Comparator) {
    Comparator["None"] = "None";
    Comparator["GreaterThan"] = ">";
    Comparator["LessThan"] = "<";
    Comparator["GreaterOrEqual"] = "\u2265";
    Comparator["LessOrEqual"] = "\u2264";
    Comparator["Equal"] = "=";
    Comparator["NotEqual"] = "\u2260";
    Comparator["Between"] = "< x <";
    Comparator["BetweenOrEqual"] = "\u2264 x \u2264";
    Comparator["BetweenOrLeftEqual"] = "\u2264 x <";
    Comparator["BetweenOrRightEqual"] = "< x \u2264";
})(Comparator || (Comparator = {}));
export const MultipleValueComparators = [
    Comparator.Between,
    Comparator.BetweenOrEqual,
    Comparator.BetweenOrLeftEqual,
    Comparator.BetweenOrRightEqual,
];
export default {};
export function isColumnMeta(column) {
    return !!column && 'column_name' in column;
}
export function isSavedExpression(column) {
    return ('column_name' in column && 'expression' in column && !!column.expression);
}
export function isControlPanelSectionConfig(section) {
    return section !== null;
}
export function isDataset(datasource) {
    return (!!datasource && 'columns' in datasource && !('sqlEditorId' in datasource));
}
export function isQueryResponse(datasource) {
    return !!datasource && 'results' in datasource && 'sqlEditorId' in datasource;
}
export var SortSeriesType;
(function (SortSeriesType) {
    SortSeriesType["Name"] = "name";
    SortSeriesType["Max"] = "max";
    SortSeriesType["Min"] = "min";
    SortSeriesType["Sum"] = "sum";
    SortSeriesType["Avg"] = "avg";
})(SortSeriesType || (SortSeriesType = {}));
