// @superset-ui/chart-controls/esm/types.js
/* eslint-disable camelcase */
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */




















// eslint-disable-next-line @typescript-eslint/no-explicit-any










/** ----------------------------------------------
 * Input data/props while rendering
 * ---------------------------------------------*/














































/**
 * The action dispatcher will call Redux `dispatch` internally and return what's
 * returned from `dispatch`, which by default is the original or another action.
 */







/**
 * Mapping of action dispatchers
 */




/**
 * Additional control props obtained from `mapStateToProps`.
 */




// Ref:superset-frontend/src/explore/store.js









// Ref: superset-frontend/src/explore/components/ControlPanelsContainer.jsx







/** ----------------------------------------------
 * Config for a chart Control
 * ---------------------------------------------*/

// Ref: superset-frontend/src/explore/components/controls/index.js
































// expanded in `expandControlConfig`

// eslint-disable-next-line @typescript-eslint/no-explicit-any




/**
 * Control config specifying how chart controls appear in the control panel, all
 * these configs will be passed to the UI component for control as props.
 *
 * - type: the control type, referencing a React component of the same name
 * - label: the label as shown in the control's header
 * - description: shown in the info tooltip of the control's header
 * - default: the default value when opening a new chart, or changing visualization type
 * - renderTrigger: a bool that defines whether the visualization should be re-rendered
 *    when changed. This should `true` for controls that only affect the rendering (client side)
 *    and don't affect the query or backend data processing as those require to re run a query
 *    and fetch the data
 * - validators: an array of functions that will receive the value of the component and
 *    should return error messages when the value is not valid. The error message gets
 *    bubbled up to the control header, section header and query panel header.
 * - warning: text shown as a tooltip on a warning icon in the control's header
 * - error: text shown as a tooltip on a error icon in the control's header
 * - shouldMapStateToProps: a function that receives the previous and current app state
 *   and determines if the control needs to recalculate it's props based on the new state.
 * - mapStateToProps: a function that receives the App's state and return an object of k/v
 *    to overwrite configuration at runtime. This is useful to alter a component based on
 *    anything external to it, like another control's value. For instance it's possible to
 *    show a warning based on the value of another component. It's also possible to bind
 *    arbitrary data from the redux store to the component this way.
 * - tabOverride: set to 'data' if you want to force a renderTrigger to show up on the `Data`
     tab, or 'customize' if you want it to show up on that tam. Otherwise sections with ALL
     `renderTrigger: true` components will show up on the `Customize` tab.
 * - visibility: a function that uses control panel props to check whether a control should
 *    be visible.
 */






























































/** --------------------------------------------
 * Additional Config for specific control Types
 * --------------------------------------------- */
















// Ref: superset-frontend/src/components/Select/SupersetStyledSelect.tsx
























/** --------------------------------------------
 * Custom controls
 * --------------------------------------------- */




// Catch-all ControlConfig
//  - if T is known control types, return SharedControlConfig,
//  - if T is object, assume a CustomComponent
//  - otherwise assume it's a custom component control









/** ===========================================================
 * Chart plugin control panel config
 * ========================================================= */




















// use ReactElement instead of ReactNode because `string`, `number`, etc. may
// interfere with other ControlSetItem types









// Ref:
//  - superset-frontend/src/explore/components/ControlPanelsContainer.jsx
//  - superset-frontend/src/explore/components/ControlPanelSection.jsx




























export const isStandardizedFormData = (
formData) => {var _formData$standardize, _formData$standardize2;return (

    (formData == null ? void 0 : (_formData$standardize = formData.standardizedFormData) == null ? void 0 : _formData$standardize.controls) && (
    formData == null ? void 0 : (_formData$standardize2 = formData.standardizedFormData) == null ? void 0 : _formData$standardize2.memorizedFormData) &&
    Array.isArray(formData.standardizedFormData.controls.metrics) &&
    Array.isArray(formData.standardizedFormData.controls.columns));};

















// Ref:
//  - superset-frontend/src/explore/components/ConditionalFormattingControl.tsx
export let Comparator = /*#__PURE__*/function (Comparator) {Comparator["None"] = "None";Comparator["GreaterThan"] = ">";Comparator["LessThan"] = "<";Comparator["GreaterOrEqual"] = "\u2265";Comparator["LessOrEqual"] = "\u2264";Comparator["Equal"] = "=";Comparator["NotEqual"] = "\u2260";Comparator["Between"] = "< x <";Comparator["BetweenOrEqual"] = "\u2264 x \u2264";Comparator["BetweenOrLeftEqual"] = "\u2264 x <";Comparator["BetweenOrRightEqual"] = "< x \u2264";return Comparator;}({});













export const MultipleValueComparators = [
Comparator.Between,
Comparator.BetweenOrEqual,
Comparator.BetweenOrLeftEqual,
Comparator.BetweenOrRightEqual];
















export default {};

export function isColumnMeta(column) {
  return !!column && 'column_name' in column;
}

export function isSavedExpression(
column)
{
  return (
    'column_name' in column && 'expression' in column && !!column.expression);

}

export function isControlPanelSectionConfig(
section)
{
  return section !== null;
}

export function isDataset(
datasource)
{
  return (
    !!datasource && 'columns' in datasource && !('sqlEditorId' in datasource));

}

export function isQueryResponse(
datasource)
{
  return !!datasource && 'results' in datasource && 'sqlEditorId' in datasource;
}

export let SortSeriesType = /*#__PURE__*/function (SortSeriesType) {SortSeriesType["Name"] = "name";SortSeriesType["Max"] = "max";SortSeriesType["Min"] = "min";SortSeriesType["Sum"] = "sum";SortSeriesType["Avg"] = "avg";return SortSeriesType;}({});
