// @boost/core/lib/CLI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class CLI {
    static registerGlobalOptions(app, tool) {
        app
            .option('config', {
            default: '',
            description: tool.msg('app:cliOptionConfig'),
            string: true,
        })
            .option('debug', {
            boolean: true,
            default: false,
            description: tool.msg('app:cliOptionDebug'),
        })
            .option('locale', {
            default: '',
            description: tool.msg('app:cliOptionLocale'),
            string: true,
        })
            .option('output', {
            default: 3,
            description: tool.msg('app:cliOptionOutput'),
            number: true,
        })
            .option('silent', {
            boolean: true,
            default: false,
            description: tool.msg('app:cliOptionSilent'),
        })
            .option('theme', {
            default: 'default',
            description: tool.msg('app:cliOptionTheme'),
            string: true,
        });
    }
}
exports.default = CLI;
